src/pcm_lowlevel_u585ai.cpp

```diff
@@ -0,0 +1,350 @@
+  DMic_OnBoard_DMAConfig.DataLength = (DEFAULT_AUDIO_IN_BUFFER_SIZE * 2U);
+  DMic_OnBoard_DMAConfig.MsbOnly    = ENABLE;
+
+  if (HAL_MDF_AcqStart_DMA(
+        &DMic_OnBoard_AdfHandle,
+        &DMic_OnBoard_AdfFilterConfig,
+        &DMic_OnBoard_DMAConfig) != HAL_OK) {
+    ret = PCM_ERROR;
+  }
+
+  if (HAL_MDF_GenerateTrgo(&DMic_OnBoard_AdfHandle) != HAL_OK) {
+    ret = PCM_ERROR;
+  }
+
+  return ret;
+}
+
+int pcm_lowlevel_stop()
+{
+  int ret = PCM_OK;
+
+  if (HAL_MDF_AcqStop_DMA(&DMic_OnBoard_AdfHandle) != HAL_OK) {
+    ret = PCM_ERROR;
+  }
+
+  return PCM_OK;
+}
+
+/**
+* @brief  Regular conversion complete callback.
+* @note   In interrupt mode, user has to read conversion value in this function
+using HAL_DFSDM_FilterGetRegularValue.
+* @param  hmdf   MDF filter handle.
+* @retval None
+*/
+void HAL_MDF_AcqCpltCallback(MDF_HandleTypeDef *hmdf)
+{
+  UNUSED(hmdf);
+
+  if (PCM.Buffer == nullptr) {
+    return;
+  }
+
+  if (DMic_OnBoard_CountSkip < 64) {
+    ++DMic_OnBoard_CountSkip;
+    return;
+  }
+
+  for (uint32_t j = 0U; j < ((PCM.SampleRate / 1000U) * N_MS_PER_INTERRUPT); j++) {
+    int32_t Z = ((DMic_OnBoard_Buffer[j + ((PCM.SampleRate / 1000U) * N_MS_PER_INTERRUPT)]) * (int32_t)(PCM.Volume)) / 100;
+    PCM.Buffer[j] = (uint16_t) SaturaLH(Z, -32760, 32760);
+  }
+
+  if (PCM.OnRecv != nullptr) {
+    PCM.OnRecv();
+  }
+}
+
+/**
+* @brief  Half regular conversion complete callback.
+* @param  hmdf   MDF filter handle.
+* @retval None
+*/
+void HAL_MDF_AcqHalfCpltCallback(MDF_HandleTypeDef *hmdf)
+{
+  UNUSED(hmdf);
+
+  if (PCM.Buffer == nullptr) {
+    return;
+  }
+
+  if (DMic_OnBoard_CountSkip < 64) {
+    ++DMic_OnBoard_CountSkip;
+    return;
+  }
+
+  for (uint32_t j = 0U; j < ((PCM.SampleRate / 1000U) * N_MS_PER_INTERRUPT); j++) {
+    int32_t Z = (DMic_OnBoard_Buffer[j] * (int32_t)(PCM.Volume)) / 100;
+    PCM.Buffer[j] = (uint16_t) SaturaLH(Z, -32760, 32760);
+  }
+
+  if (PCM.OnRecv != nullptr) {
+    PCM.OnRecv();
+  }
+}
+
+/**
+ * Initialize HAL MDF.
+ * @param  hmdf   MDF filter handle.
+ * @retval None
+ */
+void HAL_MDF_MspInit(MDF_HandleTypeDef *hmdf)
+{
+  DMA_NodeConfTypeDef dmaLinkNodeConfig;
+  GPIO_InitTypeDef  GPIO_Init;
+
+  if (hmdf->Instance == DMIC_ONBOARD_FILTER) {
+    /** Initializes the peripherals clock */
+    RCC_PeriphCLKInitTypeDef PeriphClkInit = {0};
+    PeriphClkInit.PeriphClockSelection = RCC_PERIPHCLK_ADF1;
+    PeriphClkInit.Adf1ClockSelection = RCC_ADF1CLKSOURCE_PLL3;
+    PeriphClkInit.PLL3.PLL3Source = RCC_PLLSOURCE_HSE;
+    PeriphClkInit.PLL3.PLL3M = 2;
+    PeriphClkInit.PLL3.PLL3N = 48;
+    PeriphClkInit.PLL3.PLL3P = 2;
+    PeriphClkInit.PLL3.PLL3Q = 25;
+    PeriphClkInit.PLL3.PLL3R = 2;
+    PeriphClkInit.PLL3.PLL3RGE = RCC_PLLVCIRANGE_0;
+    PeriphClkInit.PLL3.PLL3FRACN = 0;
+    PeriphClkInit.PLL3.PLL3ClockOut = RCC_PLL3_DIVQ;
+    if (HAL_RCCEx_PeriphCLKConfig(&PeriphClkInit) != HAL_OK) {
+      while (1);
```

Comment:
Would recommend to use Error_Handler();
```suggestion
      Error_Handler();
```